# Titanic

Hoje, faremos uma análise de dados do naufrágio do Titanic, um dos maiores desastres da história naval. 
A nossa principal tarefa é verificar se dado uma pessoa, caso ela estivesse no Titanic se ela teria sobrevivido ou não.
Neste tutorial, de forma bem geral, faremos a preparação dos dados de treinamento e teste, identificando dados inconsistentes ou faltantes, numéricos ou categóricos tornando assim, nosso conjunto de dados mais consistente para uma análise mais precisa. A partir disto, extrairemos correlações entre cada característica, avaliaremos nosso modelo e por fim, faremos predições.




### Algumas premissas:

* Algumas características não têm correlação com sobrevivência.

* Crianças são mais propensas a sobreviver.

* Mulheres são mais propensas a sobreviver.

* Passageiros de primeira classe têm maior probabilidade de sobreviver.

Estas 4  premissas mostraremos no decorrer deste notebook.

In [125]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier

O conjunto de dados de treino e de teste estão disponíveis no site Kaggle, porém, no diretório que este projeto está contido, também estão presentes estes dois conjuntos de dados com extensão csv.

In [126]:
treino = pd.read_csv("train.csv")
teste = pd.read_csv("test.csv")

In [127]:
treino

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
5              6         0       3   
6              7         0       1   
7              8         0       3   
8              9         1       3   
9             10         1       2   
10            11         1       3   
11            12         1       1   
12            13         0       3   
13            14         0       3   
14            15         0       3   
15            16         1       2   
16            17         0       3   
17            18         1       2   
18            19         0       3   
19            20         1       3   
20            21         0       2   
21            22         1       2   
22            23         1       3   
23            24         1       1   
24            25         0       3   
25            26         1       3   
26            27         0       3   
27            28         0       1   
28            29         1       3   
29            30         0       3   
..           ...       ...     ...   
861          862         0       2   
862          863         1       1   
863          864         0       3   
864          865         0       2   
865          866         1       2   
866          867         1       2   
867          868         0       1   
868          869         0       3   
869          870         1       3   
870          871         0       3   
871          872         1       1   
872          873         0       1   
873          874         0       3   
874          875         1       2   
875          876         1       3   
876          877         0       3   
877          878         0       3   
878          879         0       3   
879          880         1       1   
880          881         1       2   
881          882         0       3   
882          883         0       3   
883          884         0       2   
884          885         0       3   
885          886         0       3   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
5                                     Moran, Mr. James    male   NaN      0   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
7                       Palsson, Master. Gosta Leonard    male   2.0      3   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
12                      Saundercock, Mr. William Henry    male  20.0      0   
13                         Andersson, Mr. Anders Johan    male  39.0      1   
14                Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
16                                Rice, Master. Eugene    male   2.0      4   
17                        Williams, Mr. Charles Eugene    male   NaN      0   
18   Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
19                             Masselmani, Mrs. Fatima  female   NaN

In [128]:
treino.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [129]:
teste

PassengerId  Pclass                                               Name  \
0            892       3                                   Kelly, Mr. James   
1            893       3                   Wilkes, Mrs. James (Ellen Needs)   
2            894       2                          Myles, Mr. Thomas Francis   
3            895       3                                   Wirz, Mr. Albert   
4            896       3       Hirvonen, Mrs. Alexander (Helga E Lindqvist)   
5            897       3                         Svensson, Mr. Johan Cervin   
6            898       3                               Connolly, Miss. Kate   
7            899       2                       Caldwell, Mr. Albert Francis   
8            900       3          Abrahim, Mrs. Joseph (Sophie Halaut Easu)   
9            901       3                            Davies, Mr. John Samuel   
10           902       3                                   Ilieff, Mr. Ylio   
11           903       1                         Jones, Mr. Charles Cresson   
12           904       1      Snyder, Mrs. John Pillsbury (Nelle Stevenson)   
13           905       2                               Howard, Mr. Benjamin   
14           906       1  Chaffee, Mrs. Herbert Fuller (Carrie Constance...   
15           907       2      del Carlo, Mrs. Sebastiano (Argenia Genovesi)   
16           908       2                                  Keane, Mr. Daniel   
17           909       3                                  Assaf, Mr. Gerios   
18           910       3                       Ilmakangas, Miss. Ida Livija   
19           911       3              Assaf Khalil, Mrs. Mariana (Miriam")"   
20           912       1                             Rothschild, Mr. Martin   
21           913       3                          Olsen, Master. Artur Karl   
22           914       1               Flegenheim, Mrs. Alfred (Antoinette)   
23           915       1                    Williams, Mr. Richard Norris II   
24           916       1    Ryerson, Mrs. Arthur Larned (Emily Maria Borie)   
25           917       3                            Robins, Mr. Alexander A   
26           918       1                       Ostby, Miss. Helene Ragnhild   
27           919       3                                  Daher, Mr. Shedid   
28           920       1                            Brady, Mr. John Bertram   
29           921       3                                  Samaan, Mr. Elias   
..           ...     ...                                                ...   
388         1280       3                               Canavan, Mr. Patrick   
389         1281       3                        Palsson, Master. Paul Folke   
390         1282       1                         Payne, Mr. Vivian Ponsonby   
391         1283       1     Lines, Mrs. Ernest H (Elizabeth Lindsey James)   
392         1284       3                      Abbott, Master. Eugene Joseph   
393         1285       2                               Gilbert, Mr. William   
394         1286       3                           Kink-Heilmann, Mr. Anton   
395         1287       1     Smith, Mrs. Lucien Philip (Mary Eloise Hughes)   
396         1288       3                               Colbert, Mr. Patrick   
397         1289       1  Frolicher-Stehli, Mrs. Maxmillian (Margaretha ...   
398         1290       3                     Larsson-Rondberg, Mr. Edvard A   
399         1291       3                           Conlon, Mr. Thomas Henry   
400         1292       1                            Bonnell, Miss. Caroline   
401         1293       2                                    Gale, Mr. Harry   
402         1294       1                     Gibson, Miss. Dorothy Winifred   
403         1295       1                             Carrau, Mr. Jose Pedro   
404         1296       1                       Frauenthal, Mr. Isaac Gerald   
405         1297       2       Nourney, Mr. Alfred (Baron von Drachstedt")"   
406         1298       2                          Ware,

In [130]:
teste.describe()

PassengerId      Pclass         Age       SibSp       Parch        Fare
count   418.000000  418.000000  332.000000  418.000000  418.000000  417.000000
mean   1100.500000    2.265550   30.272590    0.447368    0.392344   35.627188
std     120.810458    0.841838   14.181209    0.896760    0.981429   55.907576
min     892.000000    1.000000    0.170000    0.000000    0.000000    0.000000
25%     996.250000    1.000000   21.000000    0.000000    0.000000    7.895800
50%    1100.500000    3.000000   27.000000    0.000000    0.000000   14.454200
75%    1204.750000    3.000000   39.000000    1.000000    0.000000   31.500000
max    1309.000000    3.000000   76.000000    8.000000    9.000000  512.329200


A função info() nos auxilia a verificar quantas instâncias possuem dados faltantes. Note que dados faltantes geram inconsistência no nosso dataset e afetam diretamente no nosso modelo preditivo.

In [131]:
print("Informações do conjunto de dados de treino")
print()
treino.info()

Informações do conjunto de dados de treino

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.6+ KB


Note que neste conjunto de treino possuimos 891 entradas, ou seja, instâncias do problema e 12 colunas que são os atributos. Porém, idade(age), cabine(cabin) , em qual porta embarcou(embarked) possuem dados faltantes.

In [132]:
print("Informações do conjunto de dados de teste")
teste.info()

Informações do conjunto de dados de teste
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
PassengerId    418 non-null int64
Pclass         418 non-null int64
Name           418 non-null object
Sex            418 non-null object
Age            332 non-null float64
SibSp          418 non-null int64
Parch          418 non-null int64
Ticket         418 non-null object
Fare           417 non-null float64
Cabin          91 non-null object
Embarked       418 non-null object
dtypes: float64(2), int64(4), object(5)
memory usage: 36.0+ KB


Já no conjuntos de teste temos como dados faltantes idade(age), tarifa(fare) e cabine(cabin).


### Informações sobre variáveis qualitativas.

In [133]:
# o parametro include = ['O'] apresenta apenas informações sobre
# atributos qualitativos
treino.describe(include=['O'])

Name   Sex    Ticket Cabin Embarked
count                 891   891       891   204      889
unique                891     2       681   147        3
top     Gee, Mr. Arthur H  male  CA. 2343    G6        S
freq                    1   577         7     4      644

Perceba que o sexo mais comum é masculino e a porta de embarque mais usada é a porta S.

In [134]:
teste.describe(include=['O'])

Name   Sex    Ticket            Cabin Embarked
count                            418   418       418               91      418
unique                           418     2       363               76        3
top     Collett, Mr. Sidney C Stuart  male  PC 17608  B57 B59 B63 B66        S
freq                               1   266         5                3      270

### Preparando os dados para predição 

* Preenchendo dados que não estão preenchidos.
* Correlação entre características e sobrevivência.
* Transformando variáveis categóricas em valores numéricos equivalentes.

#### 1. Remover dados que não são importantes para a análise

In [135]:
treino.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.6+ KB


* Observe que Cabine(Cabin) possui mais ausência do que presença de informação.
* Nome(Name) e PassengerId(Id do passageiro) não têm relação com sobrevivência 
* PassengerId é importante para a submissão do resultado final

In [136]:
teste_Id_passageiro = teste["PassengerId"]

treino.drop(["PassengerId","Ticket","Cabin"], axis=1, inplace=True)

teste.drop(["PassengerId","Ticket","Cabin"], axis=1, inplace=True)

treino_teste_data = [treino, teste]


#### 2. Atributo Sexo

* Não há campos em branco 
* É uma variável qualitativa

Como SEXO é uma variável qualitativa, devemos transformá-los em valores numéricos. Assim, adotamos 1 para masculino e 0 para feminino.


In [138]:
for dataset in treino_teste_data:
    dataset['Sex'] = dataset['Sex'].map({'male': 1,
                                         'female': 0}).astype(int)

In [143]:
treino_teste_data[0].head()

Survived  Pclass                                               Name  Sex  \
0         0       3                            Braund, Mr. Owen Harris    1   
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...    0   
2         1       3                             Heikkinen, Miss. Laina    0   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)    0   
4         0       3                           Allen, Mr. William Henry    1   

    Age  SibSp  Parch     Fare Embarked  
0  22.0      1      0   7.2500        S  
1  38.0      1      0  71.2833        C  
2  26.0      0      0   7.9250        S  
3  35.0      1      0  53.1000        S  
4  35.0      0      0   8.0500        S

In [144]:
treino_teste_data[1].head()

Pclass                                          Name  Sex   Age  SibSp  \
0       3                              Kelly, Mr. James    1  34.5      0   
1       3              Wilkes, Mrs. James (Ellen Needs)    0  47.0      1   
2       2                     Myles, Mr. Thomas Francis    1  62.0      0   
3       3                              Wirz, Mr. Albert    1  27.0      0   
4       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)    0  22.0      1   

   Parch     Fare Embarked  
0      0   7.8292        Q  
1      0   7.0000        S  
2      0   9.6875        Q  
3      0   8.6625        S  
4      1  12.2875        S

In [159]:
# as_index = false torna a consulta semelhante a uma consulta SQL.
# teste com valor as_index = True
treino[['Sex','Survived']].groupby(['Sex'],
                                   as_index = False).mean().sort_values(by = 'Survived', ascending = True)

Sex  Survived
1    1  0.188908
0    0  0.742038

#### Com esta consulta, podemos considerar que pessoas do sexo feminino são mais propensos a sobreviver.

#### Atributo Pclass
Este atributo indica qual classe cada indivíduo está podendo ser 1ª, 2ª ou 3ª classe.

In [165]:
treino[['Pclass', 'Survived']].groupby(['Pclass'], 
                                        as_index = False).mean().sort_values(by = 'Survived', ascending = True)

Pclass  Survived
2       3  0.242363
1       2  0.472826
0       1  0.629630

Desta consulta, podemos concluir que passageiros da 1ª classe estão mais propensos e passageiros da 3ª menos e os de 2ª classe está no meio destas duas classes.